app/controllers/concerns/current_role.rb

```diff
@@ -0,0 +1,43 @@
+# CurrentRole adds support allowing the requestor to assume another
+# role. There are two ways to specify the role that should be
+# assumed. If the Conjur-Assume-Role header is present on the request,
+# it will be used. Alternatively, a query paramter (:acting_as by
+# default) can be used. If neither the header nor the query paramter
+# is present, current_role will be current_user.
+
+# The authenticated user must hold that role, or
+# ApplicationController::Forbidden will be raised.
+
+module CurrentRole
+  extend ActiveSupport::Concern
+
+  def current_role?
+    begin
+      current_role
+    rescue Forbidden
+      nil
+    end
+  end
+
+  def current_role
+    @current_role ||= find_current_role(:acting_as)
+  end
+
+  def current_role_from_param(param_name)
+    find_current_role(param_name)
+  end
+  
+  private
+
+  def find_current_role(param_name)
+    acting_as = request.headers['Conjur-Assume-Role'].presence || params[param_name].presence
+    return current_user unless acting_as
+
+    role = Role[acting_as]
+    raise ApplicationController::Forbidden unless role && role.is_ancestor_of?(current_user)
```

Comment:
Raising `Forbidden` if the requested role doesn't exist seems a little weird. Maybe both cases should be 404, indicating that the role you requested doesn't exist, or maybe it does exist but you don't have access to it.

Reply:
Roles aren't currently visibility filtered (only resources are), so returning 404 on non-held role would be a bit weird.

For a nonexistent role 404 would be a wrong response for a header -- because 404 semantically says the the target resource (in REST sense) does not exist, which wouldn't be true. For a query parameter, I suppose returning 404 does make some sense in case the role doesn't exist.